### PR TITLE
ncm-metaconfig: add service ssh

### PR DESCRIPTION
**Pull request by wdpypere**
allows for ssh client config to be written.

### CODE REVIEW

File: ncm-metaconfig/src/main/metaconfig/ssh/client.tt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+[% INCLUDE metaconfig/ssh/client_attrs.tt data=main -%]
+
+[% FOREACH mt IN Match -%]
+Match [%    mt.matches.join(' ') %]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client.tt]: no indent before the `mt`

File: ncm-metaconfig/src/main/metaconfig/ssh/client.tt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+[% INCLUDE metaconfig/ssh/client_attrs.tt data=mt FILTER indent %]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client.tt]: indent `INCLUDE`

File: ncm-metaconfig/src/main/metaconfig/ssh/client.tt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+[% END -%]
+
+[% FOREACH hd IN Host -%]
+Host [%    hd.hostnames.join(' ') %]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client.tt]: no indent before `hd`

File: ncm-metaconfig/src/main/metaconfig/ssh/client.tt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+[% INCLUDE metaconfig/ssh/client_attrs.tt data=hd FILTER indent %]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client.tt]: indent `INCLUDE`

File: ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+[% spacelist = ['SendEnv', 'GlobalKnownHostsFile', 'IgnoreUnknown', 'UserKnownHostsFile' ] -%]
+[% commalist = ['Ciphers', 'HostbasedKeyTypes','HostKeyAlgorithms', 'KbdInteractiveDevices', 'KbdInteractiveDevices',
+    'MACs', 'PreferredAuthentications', 'CanonicalDomains', 'CanonicalizePermittedCNAMEs', 'KexAlgorithms'] -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt]: end multiline lists with a `,` and the `] -%]` on a new line. will make the diff easier when adding new keys.

File: ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+[% multilinelist = ['IdentityFile', 'RevokedHostKeys'] -%]
+[% booleans = ['BatchMode', 'CanonicalizeFallbackLocal', 'ChallengeResponseAuthentication', 'CheckHostIP', 
+    'ClearAllForwardings', 'Compression', 'EnableSSHKeysign', 'ExitOnForwardFailure', 'ForwardAgent', 'ForwardX11', 
+    'ForwardX11Trusted', 'GatewayPorts', 'GSSAPIAuthentication', 'GSSAPIDelegateCredentials', 'HashKnownHosts', 
+    'HostbasedAuthentication', 'IdentitiesOnly', 'KbdInteractiveAuthentication', 'NoHostAuthenticationForLocalhost', 
+    'PasswordAuthentication', 'PermitLocalCommand', 'ProxyUseFdpass', 'PubkeyAuthentication', 'RhostsRSAAuthentication', 
+    'RSAAuthentication', 'StreamLocalBindUnlink', 'TCPKeepAlive', 'UsePrivilegedPort', 'VisualHostKey'] -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt]: same remark

File: ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+[%- FOREACH pair IN data.pairs -%]
+[% NEXT IF pair.key == 'hostnames' || pair.key == 'matches' -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt]: indent `NEXT`

File: ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+[% NEXT IF pair.key == 'hostnames' || pair.key == 'matches' -%] 
+[%      SWITCH pair.key -%]
+[%          CASE booleans -%]
+[%              pair.key %] [% pair.value ? 'Yes' : 'No' %]
+[%          CASE spacelist -%]
+[%              pair.key %] [% pair.value.join(' ') %]
+[%          CASE commalist -%]
+[%              pair.key %] [% pair.value.join(',') %]
+[%          CASE multilinelist -%]
+[%             FOREACH line IN pair.value -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt]: 4 space indent

File: ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+[%                  pair.key %] [% line %]
+[%             END -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/client_attrs.tt]: also here 4 space indent

File: ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan [hunk 10]
```diff
@@ -0,0 +1,115 @@
+declaration template metaconfig/ssh/schema;
+
+include 'pan/types';
+
+type ssh_ciphers = string with match (SELF, "^(3des-cbc|aes128-cbc|aes192-cbc|aes256-cbc|aes128-ctr|aes192-ctr|aes256-ctr|aes128-gcm@openssh.com|aes256-gcm@openssh.com|arcfour|arcfour128|arcfour256|blowfish-cbc|cast128-cbc|chacha20-poly1305@openssh.com)$"); 
+type ssh_hostkeyalgorithms = string with match(SELF, "^(ssh-rsa|ssh-dss|ssh-ed25519|ecdsa-sha2-nistp256|ecdsa-sha2-nistp384|ecdsa-sha2-nistp521|ssh-rsa-cert-v01@openssh.com|ssh-dss-cert-v01@openssh.com|ecdsa-sha2-nistp256-cert-v01@openssh.com|ecdsa-sha2-nistp384-cert-v01@openssh.com|ecdsa-sha2-nistp521-cert-v01@openssh.com|ssh-rsa-cert-v00@openssh.com|ssh-dss-cert-v00@openssh.com|ssh-ed25519-cert-v01@openssh.com)$");
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: group the `@openssh.com` common strings to shorten the line

Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: same for line below, try to group common words in the regexps

File: ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+type ssh_kbdinteractivedevices = string with match (SELF, "^(bsdauth|pam|skey)$");
+type ssh_kexalgorithms = string with match (SELF, "^(diffie-hellman-group1-sha1|diffie-hellman-group14-sha1|diffie-hellman-group-exchange-sha1|diffie-hellman-group-exchange-sha256|ecdh-sha2-nistp256|ecdh-sha2-nistp384|ecdh-sha2-nistp521|diffie-hellman-group1-sha1|curve25519-sha256@libssh.org|gss-gex-sha1-|gss-group1-sha1-|gss-group14-sha1-)$");
+type ssh_MACs = string with match(SELF, "^(hmac-sha1|hmac-sha1-96|hmac-sha2-256|hmac-sha2-512|hmac-md5|hmac-md5-96|hmac-ripemd160|hmac-ripemd160@openssh.com|umac-64@openssh.com|umac-128@openssh.com|hmac-sha1-etm@openssh.com|hmac-sha1-96-etm@openssh.com|hmac-sha2-256-etm@openssh.com|hmac-sha2-512-etm@openssh.com|hmac-md5-etm@openssh.com|hmac-md5-96-etm@openssh.com|hmac-ripemd160-etm@openssh.com|umac-64-etm@openssh.com|umac-128-etm@openssh.com)$");
+
+
+type ssh_config_opts = {
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: if this is generated with some magic one liner, maybe paste it here?

File: ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    'HashKnownHosts' ? boolean
+    'HostbasedAuthentication' ? boolean
+    'HostbasedKeyTypes' ? string[]
+    'HostKeyAlgorithms' ? ssh_hostkeyalgorithms[] 
+    'HostKeyAlias' ? string
+    'HostName' ? string
+    'IdentitiesOnly' ? boolean
+    'IdentityFile' ? string[]
+    'IgnoreUnknown' ? string[]
+    'IPQoS' ? string with match (SELF, "^(af11|af12|af13|af21|af22|af23|af31|af32|af33|af41|af42|af43|cs0|cs1|cs2|cs3|cs4|cs5|cs6|cs7|ef|lowdelay|throughput|reliability)$")
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: group common regexps like `af[1234][123]` for all `terms`

File: ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    'KbdInteractiveAuthentication' ? boolean
+    'KbdInteractiveDevices' ? ssh_kbdinteractivedevices[]
+    'KexAlgorithms' ?  ssh_kexalgorithms[]
+    'LocalCommand' ? string
+    'LocalForward' ? string
+    'LogLevel' ? string with match (SELF, "^(QUIET|FATAL|ERROR|INFO|VERBOSE|DEBUG|DEBUG1|DEBUG2|DEBUG3)$")
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: `DEBUG[123]?`

File: ncm-metaconfig/src/main/metaconfig/ssh/tests/profiles/allopts.pan [hunk 14]
```diff
@@ -0,0 +1,88 @@
+object template allopts;
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/tests/profiles/allopts.pan]: rename the template to `client_allopts` (while awaintg your PR for `server_allopts` :smile:

File: ncm-metaconfig/src/main/metaconfig/ssh/tests/profiles/config.pan [hunk 15]
```diff
@@ -0,0 +1,51 @@
+object template config;
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/tests/profiles/config.pan]: rename template to `client_config`

File: ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/allopts/base [hunk 16]
```diff
@@ -0,0 +1,88 @@
+Base test for all ssh client config options
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/allopts/base]: rename subdir to `clinet_allopts`

File: ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/allopts/base [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+---
+/etc/ssh/ssh_config
+---
+^AddressFamily\sany$
+^BatchMode\sYes$
+^BindAddress\s192.168.1.1$
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/allopts/base]: the `.` match any character, so you might want to escape them

File: ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/config/base [hunk 18]
```diff
@@ -0,0 +1,28 @@
+Base test for ssh client config
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/tests/regexps/config/base]: rename subdir to `client_config`

**Comment by stdweird** [comment]
@wdpypere just some minor remark. try to make the regexp a bit more compact; otherwise really nice work!

**Comment by wdpypere** [comment]
@stdweird remarks addressed.

File: ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan [hunk 19]
```diff
@@ -0,0 +1,115 @@
+declaration template metaconfig/ssh/schema;
+
+include 'pan/types';
+
+type ssh_ciphers = string with match (SELF, "^((blowfish|3des|aes128|aes192|aes256|cast128)-cbc|(aes128|aes192|aes256)-ctr|arcfour|arcfour(128|256)|(aes128-gcm|aes256-gcm|chacha20-poly1305)@openssh.com)$");
+type ssh_hostkeyalgorithms = string with match(SELF, "^(ssh-rsa|ssh-dss|ssh-ed25519|ecdsa-sha2-nistp256|ecdsa-sha2-nistp384|ecdsa-sha2-nistp521|ssh-rsa-cert-v01@openssh.com|ssh-dss-cert-v01@openssh.com|ecdsa-sha2-nistp256-cert-v01@openssh.com|ecdsa-sha2-nistp384-cert-v01@openssh.com|ecdsa-sha2-nistp521-cert-v01@openssh.com|ssh-rsa-cert-v00@openssh.com|ssh-dss-cert-v00@openssh.com|ssh-ed25519-cert-v01@openssh.com)$");
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/ssh/pan/schema.pan]: more `@openssh.com` to group

**Comment by stdweird** [comment]
LGTM!

**Pull request merged**